lib/Sema/CSSimplify.cpp

```diff
@@ -4225,52 +4225,25 @@ static bool repairArrayLiteralUsedAsDictionary(
     ConstraintKind matchKind,
     SmallVectorImpl<RestrictionOrFix> &conversionsOrFixes,
     ConstraintLocator *loc) {
+  if (auto *fix = TreatArrayLiteralAsDictionary::attempt(cs, dictType,
+                                                         arrayType, loc)) {
+    // Ignore any attempts at promoting the value to an optional as even after
+    // stripping off all optionals above the underlying types don't match (array
+    // vs dictionary).
+    conversionsOrFixes.erase(
+        llvm::remove_if(conversionsOrFixes,
+                        [&](RestrictionOrFix &E) {
+                          if (auto restriction = E.getRestriction())
+                            return *restriction ==
+                                   ConversionRestrictionKind::ValueToOptional;
```

Comment:
Just adjusted! Had to do a minor adjustment as well to move the DeepEquality restriction check to before recording a fix for OptionalToOptional, because that was the source of failing diagnostic in the `m["a"] = Optional(["", ""])` case.